commands/operator-sdk/cmd/generate/k8s.go

```diff
@@ -50,71 +51,148 @@ func k8sFunc(cmd *cobra.Command, args []string) {
 	K8sCodegen()
 }
 
+const (
+	// k8sVerTag is the k8s.io/code-generator tag used to build codegen binaries.
+	k8sVerTag = "kubernetes-1.11.2"
+	// codegenRepo is the git repo path for k8s.io/code-generator.
+	codegenRepo = "https://github.com/kubernetes/code-generator.git"
+)
+
 // K8sCodegen performs deepcopy code-generation for all custom resources under pkg/apis
 func K8sCodegen() {
-
 	projutil.MustInProjectRoot()
+
 	repoPkg := projutil.CheckAndGetProjectGoPkg()
-	outputPkg := filepath.Join(repoPkg, "pkg/generated")
-	apisPkg := filepath.Join(repoPkg, scaffold.ApisDir)
-	groupVersions, err := parseGroupVersions()
+	codegenVendor := filepath.Join("vendor", "k8s.io", "code-generator")
+
+	if _, err := os.Stat(codegenVendor); err != nil {
+		cloneCodegenRepo(codegenVendor, codegenRepo, k8sVerTag)
+	}
+	installCodegenBinaries(codegenVendor)
+
+	gvMap, err := parseGroupVersions()
 	if err != nil {
 		log.Fatalf("failed to parse group versions: (%v)", err)
 	}
+	gvStr := ""
+	for g, vs := range gvMap {
+		gvStr += fmt.Sprintf("%s:%v, ", g, vs)
+	}
 
-	log.Infof("Running code-generation for Custom Resource group versions: [%s]\n", groupVersions)
+	log.Infof("Running code-generation for Custom Resource group versions: [%v]\n", gvStr)
 
-	// TODO: Replace generate-groups.sh by building the vendored generators(deepcopy, lister etc)
-	// and running them directly
-	genGroupsCmd := "vendor/k8s.io/code-generator/generate-groups.sh"
-	args := []string{
-		"deepcopy",
-		outputPkg,
-		apisPkg,
-		groupVersions,
-	}
-	cgCmd := exec.Command(genGroupsCmd, args...)
-	cgCmd.Stdout = os.Stdout
-	cgCmd.Stderr = os.Stderr
-	err = cgCmd.Run()
-	if err != nil {
-		log.Fatalf("failed to perform code-generation: (%v)", err)
-	}
+	deepcopyGen(repoPkg, gvMap)
 
 	log.Info("Code-generation complete.")
 }
 
-// getGroupVersions parses the layout of pkg/apis to return the API groups and versions
-// in the format "groupA:v1,v2 groupB:v1 groupC:v2",
-// as required by the generate-groups.sh script
-func parseGroupVersions() (string, error) {
-	var groupVersions string
+func cloneCodegenRepo(installPath, repo, tag string) {
+	if err := os.RemoveAll(installPath); err != nil {
+		log.Fatal(err)
+	}
+	cloneCmd := exec.Command("git", "clone", "-q", repo, installPath)
+	if err := projutil.ExecCmd(cloneCmd); err != nil {
+		log.Fatal(err)
+	}
+	checkoutCmd := exec.Command("git", "checkout", "-q", tag)
+	checkoutCmd.Dir = installPath
+	if err := projutil.ExecCmd(checkoutCmd); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func installCodegenBinaries(installPath string) {
+	args := []string{
+		"install",
```

Comment:
I don't think we should use `install`, as that installs it for the user in their `$GOPATH/bin`, overwriting copies/versions of generators they may have installed themselves. Even though that is the way that the code-generator repo's scripts worked, I think a better idea if we could find a way to use `build` and output to a local directory instead of using `install`.

Reply:
I agree. @hasbro17 thoughts?